exporter/exporterhelper/metrics_batch.go

```diff
@@ -0,0 +1,236 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package exporterhelper // import "go.opentelemetry.io/collector/exporter/exporterhelper"
+
+import (
+	"context"
+	"errors"
+
+	"go.opentelemetry.io/collector/exporter/exporterbatcher"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+)
+
+// mergeMetrics merges two metrics requests into one.
+func mergeMetrics(_ context.Context, r1 Request, r2 Request) (Request, error) {
+	mr1, ok1 := r1.(*metricsRequest)
+	mr2, ok2 := r2.(*metricsRequest)
+	if !ok1 || !ok2 {
+		return nil, errors.New("invalid input type")
+	}
+	mr2.md.ResourceMetrics().MoveAndAppendTo(mr1.md.ResourceMetrics())
+	return mr1, nil
+}
+
+// mergeSplitMetrics splits and/or merges the metrics into multiple requests based on the MaxSizeConfig.
+func mergeSplitMetrics(_ context.Context, cfg exporterbatcher.MaxSizeConfig, r1 Request, r2 Request) ([]Request, error) {
+	var (
+		res          []Request
+		destReq      *metricsRequest
+		capacityLeft = cfg.MaxSizeItems
+	)
+	for _, req := range []Request{r1, r2} {
+		if req == nil {
+			continue
+		}
+		srcReq, ok := req.(*metricsRequest)
+		if !ok {
+			return nil, errors.New("invalid input type")
+		}
+		if srcReq.md.DataPointCount() <= capacityLeft {
+			if destReq == nil {
+				destReq = srcReq
+			} else {
+				srcReq.md.ResourceMetrics().MoveAndAppendTo(destReq.md.ResourceMetrics())
+			}
+			capacityLeft = cfg.MaxSizeItems - destReq.md.DataPointCount()
+			continue
+		}
+
+		for {
+			extractedMetrics := extractMetrics(srcReq.md, capacityLeft)
+			if extractedMetrics.DataPointCount() == 0 {
+				break
+			}
+			capacityLeft -= extractedMetrics.DataPointCount()
+			if destReq == nil {
+				destReq = &metricsRequest{md: extractedMetrics, pusher: srcReq.pusher}
+			} else {
+				extractedMetrics.ResourceMetrics().MoveAndAppendTo(destReq.md.ResourceMetrics())
+			}
+
+			if capacityLeft <= 0 {
```

Comment:
I don't understand when `capacityLeft` would be negative. Could you add a comment?

Reply:
It cannot be. I can change it to `==` to make it clear. We can rely on tests instead of adding extra safeguards in the production code